src/librbd/mirror/snapshot/CreatePrimaryRequest.cc

```diff
@@ -177,13 +177,46 @@ void CreatePrimaryRequest<I>::handle_refresh_image(int r) {
 
 template <typename I>
 void CreatePrimaryRequest<I>::unlink_peer() {
+  // TODO: Document semantics for unlink_peer
   uint64_t max_snapshots = m_image_ctx->config.template get_val<uint64_t>(
     "rbd_mirroring_max_mirroring_snapshots");
   ceph_assert(max_snapshots >= 3);
 
   std::string peer_uuid;
   uint64_t snap_id = CEPH_NOSNAP;
 
+  // If snap has no peers, remove snapshot. This can happen when mirror
+  // removes peer from peer_uuids when sync is completed and no other
+  // peers remain.
+  for (auto &peer : m_mirror_peer_uuids) {
+    std::shared_lock image_locker{m_image_ctx->image_lock};
+    for (auto &snap_it : m_image_ctx->snap_info) {
+      auto info = std::get_if<cls::rbd::MirrorSnapshotNamespace>(
+        &snap_it.second.snap_namespace);
+      if (info == nullptr) {
+        continue;
+      }
+      if (info->mirror_peer_uuids.empty() || !info->complete) {
```

Comment:
Need to ensure snapshot is not linked to other peer(s) before acting on `!info->complete`:
```
      if (!info->mirror_peer_uuids.empty() &&
          info->mirror_peer_uuids.count(peer) == 0) {
        // snapshot is linked to other peer(s)
        continue;
      }
```